core/api.go

```diff
@@ -28,8 +28,55 @@ func (core *Core) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		core.initJS(w, req)
 		return
 
-	case match(req, `GET /auth/oauth/.+`):
-		core.oauth2(w, req)
+	case match(req, `GET /auth/([^/]+)/(redir|web|cli)`):
+		re := regexp.MustCompile("/auth/([^/]+)/(redir|web|cli)")
+		m := re.FindStringSubmatch(req.URL.Path)
+
+		name := m[1]
+		provider, err := core.FindAuthProvider(name)
+		if err != nil {
+			w.WriteHeader(404)
+			fmt.Fprintf(w, "couldn't find that auth provider...: %s", err)
+			return
+		}
+
+		if m[2] == "redir" {
+			via := "web"
+			if cookie, err := req.Cookie("via"); err == nil {
+				via = cookie.Value
+			}
```

Comment:
Should we kill the via Cookie at this point

Reply:
Thought about it, but it doesn't hurt anything, and it makes it easier to troubleshoot ex post facto.